### ISSUE TITLE
Add support for remotely fetching and updating DataSets from JSON sources

### ISSUE THREAD

**Issue by dasgarner** [xibosignage/xibo#1320]
https://github.com/xibosignage/xibo-cms/pull/283

As I understand it, from reviewing the PR submitted by @LukyLuke the purpose of this enhancement is to add a new type of column to a DataSet which is considered a remote column.

Data for this remote column is automatically fetched by a Task and kept in sync with a remote JSON datasource (URL, etc).

------

To concretise:
The new RemoteDatasets allows you to fetch and truncate Datasets based on JSON-Requests to other systems. The Remote-Calls are done with the new Task RemoteDataSetFetchTask.
The new Entity and Controller is based on the already existing ones from the DataSet to use all already existing functions based on them.

What's missing:

Documentation :o)
I am not aware of this PHP-Doc styles and tags
Unit Tests (I've never worked with PHPUnit but with JUnit)
For the License: I hope I forgot nothing, I also can change all to xibo signange, just ping me for that.

NB: Don't be confused about the composer.phar stuff, that was added by mistake and reverted.
NB: This work I've done for change my xibo_rgraph module to only use DataSets and no direct JSON Calls (what makes problems due to (fucking) CORS)

**Comment by dasgarner** [comment]
LukyLuke's initial work has been merged into feature/remote-datasets so that it can be worked on in a collaborative fashion.

**Comment by dasgarner** [comment]
@LukyLuke 

So I have pulled your branch into my dev env and got everything running - the forms, etc seem to work well.

I have lots of questions - but from my first review of the work my main curiosities are:

### Do you have an example use case I can look at to try and help me understand how everything works?
I do sort of understand, but I must admit i'm grasping at the edges without a sample JSON and config. 

### How to the dependents work
I've tried to work it out, but can't :smile: if the above example could contain dependents, that would be great.

If there are dependencies, then does the Layout import/export need to be aware of those dependencies? For that matter does the Layout import/export need to be aware of Remote DataSets?

### What pushed you to create a new table, rather than extending the `dataset` table with new fields?
I understand that those fields would need to be nullable or have default values, but they all do anyway - and having one table would make the logic easier to understand. You'd have one controller too and one form, etc, etc.

It also prevents people doing stupid things - such as using the `DELETE /dataset/:id` route on a remote DataSet.

This seems to be particularly relevant as you have a new DataSetColumnType - and again there wouldn't be anything to prevent a user adding one to a normal DataSet.

I think my overall feeling is that we should combine the concepts - i'm not asking you to make the changes, I can do that no problem!

### Have you handled notify?
When you edit a "normal" DataSet it notifies Layouts -> Displays that the data has changed. I'm not sure if you do that?

### Processing
You have some processing logic in the controller - I think this is only ever called from the Task to update the results? If so I might be inclined to move those into the task so its obvious what scope they occupy.


There are a bunch of other little tweaks and such - with your permission I will make some of those on my side before handing it back over to you for your analysis?

**Comment by LukyLuke** [comment]
Hey @dasgarner, 
nice, thanks for your review and questions!

First of all: The need for this came by our Dev-Teams to display the amount of criticals/blockers etc. mesasured by Sonarcube, displays the amount of incomming vs. closing Bugs per day (Jira), UnitTest Coverage (sonar) and Build-Processes from Jenkins...

## New table instead reusing `dataset` table
I did this exactly for the reason to seperate the Attributtes. I separated the Classes (Entities and so) so I also separated the tables. But you're right, haveing them together would be somehow nicer but also end up in a bigger controller (as you mentioned). I can combine this, if you more preffer it that way. I am completely free of emotions on this 😄 

## Notify on change
No, I guess I did not. But Probably it's done already by your existing logic due to I just extended the Entities and DataSet?

## Processing
Good point. Actually I was thinking of a "Refetch on every visit" option, but I am not sure if there is a need for this. Probably it's better to fetch it every 5 Minutes or so 🤔 

## Dependencies
Do you mean the dependencies from the Code/Layout side? Or the "depends on" option on the RemoteDataSet Settings?

The Dependencies on the Layout-Side: I tried it to make the layout to depend on the Dataset-Layout - But I couldn't figure out how this is possible.

The *Depends on* Setting: Let's say you have a DataSource where you can fetch all Projects you're working on. These you store in a RemoteDataSet (column: ProjectName, Option: Clear on every Request) and use them in a second one to fetch statistics about bugs or whatever. So on the second one you can use `{{COL.ProjectName}}` in the URI or Post-Data (also `{{DATE}}` or `{{TIME}}` for example to fetch something for the current date and/or time).

## Examples
### Jira
Get all Bugs from today to display these in a Graph wo on a simple table.
+Call:+ https://jira.example.com/rest/api/2/search?jql=Team%20in%20(SuperDuperTeam)%20AND%20issuetype%20%3D%20Bug%20AND%20status%20was%20in%20(%22Bereit%20zur%20Umsetzung%22%2C%20Entwicklung%2C%20QA)%20on%20{{DATE}}&fields=id,priority,customfield_12910,customfield_10510
+Result:+ [jira_issues.json](https://github.com/xibosignage/xibo/files/1527585/jira_issues.json.txt)

* All the Issues are in the field `issues`, therefore use this as *Root Path* on the RemoteDataSet
* Fields: priority,customfield_12910,customfield_10510
** DataColum Priority: `fields.priority.name`
** DataColum Customer: `fields.customfield_12910.value`
** DataColum Component: `fields.customfield_10510.value`

If you are only interrested in the amout of "high", "middle" and so on Bugs, you can use only the Priority field and use the "Count" Method on a second Column where you count the `field.priority.id`...

### Sonar Cube
Get statistics about the Code-Coverage from SonarCube
+Call:+ http://sonarcube.example.com/api/timemachine?resource=project-name&metrics=coverage,it_coverage,line_coverage,branch_coverage
+Result:+ [sonar_coverage.json](https://github.com/xibosignage/xibo/files/1527586/sonar_coverage.json.txt)

* The interresting values are in the `0.cells`, spo set this as Root in the RemoteDataSet
* Values are under `v.0`, `v.1`, `v.2`

### Jenkins
For example to display which build has been successfull or not, to count the the numbers or whatever 🤔 
+Call:+ http://ci.example.com/job/pipeline/job/name/api/json?tree=builds[building,duration,estimatedDuration,result,displayName]
+Result:+ [jenkins_example.json](https://github.com/xibosignage/xibo/files/1527587/jenkins_example.json.txt)

* The Root here is `builds`
* The Fields `duration`, `displayName` and `result`
** The Field `result` may also be used for the Count function


# PullRequest Failure:
Oh fu..., I just found an ugly error in my PullRequest:
`lib/Middleware/State.php:603` -> `/*IconDa*/shboard` should be `IconDashboard`

**Comment by dasgarner** [comment]
Thank you for providing such concrete examples - I will try those out today and get back to you!

I've been considering the old table vs new table discussion and I have a preference for keeping it in 1 table. Overall I think this will be clearer to understand for maintenance and troubleshooting.

I am happy to make the necessary changes - or would you like to do so and resubmit a new PR?

The dependencies look really interesting - having the ability to link data sources like that could be very powerful!

**Comment by LukyLuke** [comment]
I am completely with ou, yes. After thinking more about it it's more clear in one Table.
Please feel free to change anything you like! :)

**Comment by dasgarner** [comment]
I've started making some alterations, which you're more than welcome to review (you can pull from my branch).

I've combined the tables and entities, as well as moving some things around to places that made more sense to me (for example the factory being responsible for processing results, rather than the controller - meaning the task doesn't need access to the controller).

I've also switched native cURL out for Guzzle, which allowed some simplification to the way parameters are substituted.

And added some logging as i've explored what's happening.

I still need to verify everything still works, particularly with dependant DataSets (in a chain) and adjust the API docs and manual. I also want to simplify the form so that if "Remote?" isn't ticked you don't see the other tabs.

Its all looking very neat though - i've got some of my own examples in mind, which I will try out once the above is done.

Any comments are most welcome!
